include/pressio/ops/ops_ordinal_type.hpp

```diff
@@ -0,0 +1,158 @@
+// POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Francesco Rizzi (fnrizzi@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef OPS_OPS_ORDINAL_TYPE_HPP_
+#define OPS_OPS_ORDINAL_TYPE_HPP_
+
+namespace pressio{ namespace ops{ namespace impl{
+
+// ContainerInfo extracts types from containers and expressions in a unified way
+// - without the expressions defining same things as the original containers
+// (e.g. StorageIndex in Eigen expr)
+// Note: it's package-specific and should be used after type resolution
+template<typename T, class Enable = void> struct ContainerInfo;
+
+#ifdef PRESSIO_ENABLE_TPL_EIGEN
+template <class T>
+struct ContainerInfo<
+  T,
+  ::pressio::mpl::enable_if_t<
+    ::pressio::is_vector_eigen<T>::value
+    || ::pressio::is_dense_matrix_eigen<T>::value
+    >
+  >
+{
+  using ordinal_type = typename T::StorageIndex;
+};
+
+// following definitions consider native Eigen expression types
+template <class VectorType, int Size>
+struct ContainerInfo<Eigen::VectorBlock<VectorType, Size>>
+: public ContainerInfo<VectorType>
+{};
+
+template <class MatrixType, int DiagIndex>
+struct ContainerInfo<Eigen::Diagonal<MatrixType, DiagIndex>>
+: public ContainerInfo<MatrixType>
+{};
+
+template<class XprType, int BlockRows, int BlockCols, bool InnerPanel>
+struct ContainerInfo<Eigen::Block<XprType, BlockRows, BlockCols, InnerPanel>>
+: public ContainerInfo<XprType>
+{};
+
+#endif
+
+#ifdef PRESSIO_ENABLE_TPL_KOKKOS
+template <class T>
+struct ContainerInfo<
+  T,
+  ::pressio::mpl::enable_if_t<
+    ::pressio::is_vector_kokkos<T>::value
+    || ::pressio::is_dense_matrix_kokkos<T>::value
+    >
+  >
+{
+  using ordinal_type = typename T::traits::size_type;
+};
+#endif
+
+#ifdef PRESSIO_ENABLE_TPL_TRILINOS
+template <class T>
+struct ContainerInfo<
+  T,
+  ::pressio::mpl::enable_if_t<
+    ::pressio::is_vector_tpetra<T>::value
+    || ::pressio::is_vector_tpetra_block<T>::value
+    || ::pressio::is_multi_vector_tpetra_block<T>::value
+    || ::pressio::is_multi_vector_tpetra<T>::value
+    >
+ >
+{
+  using global_ordinal_type = typename T::global_ordinal_type;
+  using local_ordinal_type = typename T::local_ordinal_type;
+};
+
+template <class T>
+struct ContainerInfo<
+  T,
+  ::pressio::mpl::enable_if_t<
+    ::pressio::is_vector_epetra<T>::value
+    || ::pressio::is_multi_vector_epetra<T>::value
+    >
+ >
+{
+  using global_ordinal_type = int;
+  using local_ordinal_type = int;
+};
+#endif
+
+template <class T>
+struct ContainerInfo<
+  T,
+  ::pressio::mpl::enable_if_t<
+    ::pressio::is_expression<T>::value
+    >
+  >
+{
+private:
+  using native_type = typename ::pressio::Traits<T>::native_expr_type;
+public:
+  using ordinal_type = typename ContainerInfo<native_type>::ordinal_type;
+};
+
+// short predicate-like form
+template <class T>
+using ordinal_type = typename ContainerInfo<T>::ordinal_type;
```

Comment:
I am not in favor of this syntax because this diverges from the typical way to get a type in std which relies on the ```_t``` notation. Usually, one sees things like ```std::remove_cvref_t`` to get the nested type. 
Also, there is no advantage in having a ContainerInfo. 

So I would like please to change this so that we can do this: 
```
tempalte<class T, class = void>
struct OrdinalType{
  using type = void;
};

tempalte<class T, class = void>
struct OrdinalType<
  T, ::pressio::mpl::enable_if_t<::pressio::is_expression<T>::value>
  >
{
private:
  using native_type = typename ::pressio::Traits<T>::native_expr_type;
public:
  using type = typename ContainerInfo<native_type>::ordinal_type;
};

template <class T>
using ordinal_type_t = typename OrdinalType<T>::type;
```

And similarly for local ordinal and global ordinal.